### PR TITLE
Updaters

### PR DESCRIPTION
**Pull request by bsuwal**
This PR got more bloated than I intended to. Some things that I am particularly wonky about:
* I chose to rename `updaters` as `measures` . I started with Measures of Interest and settled for `measures` but I am not quite happy with it, if you can think of a better name then let me know.
* The measures are stored with the first measure storing everything and each subsequent one storing just the changes

I have also commented on the PR to explain any weird non - updaters code.

### CODE REVIEW

File: src/graph.jl [hunk 1]
```diff
@@ -1,7 +1,3 @@
-# using JSON
-# using SparseArrays
-# using LightGraphs
-
```
Comment (bsuwal) [inline src/graph.jl]: Just some cleanups

File: src/measures.jl [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+    # we don't want to alter the data in all_measures
+    measures = deepcopy(all_measures[1])
+
+    for i in 2:step
+        curr_measures = all_measures[i]
+        (D₁, D₂) = all_measures[i]["dists"]
+
+        for key in keys(measures)
+            # TODO: this handling for non-array singular values is poor
```
Comment (bsuwal) [inline src/measures.jl]: Leaving this TODO here for now, will come back to this when I (or you) implement partisan metrics and we have more of these values that are of length 1 (like efficiency gap)

File: src/partition.jl [hunk 3]
```diff
@@ -133,3 +111,24 @@ function sample_adjacent_districts_randomly(partition::Partition, num_dists::Int
...
+    for i in 1:num_nodes
+        populations[i] = graph[nodes_str][i][pop_col]
+        if graph[nodes_str][i][assignment_col] isa String
+            assignments[i] = parse(Int, graph[nodes_str][i][assignment_col])
+        elseif graph[nodes_str][i][assignment_col] isa Int
+            assignments[i] = graph[nodes_str][i][assignment_col]
+        end
+    end
+    return populations, assignments
+end
```
Comment (bsuwal) [inline src/partition.jl]: Got confused and moved the function around, dont mind this.

File: src/recom.jl [hunk 4]
```diff
@@ -83,29 +83,6 @@ function traverse_mst(mst::Dict{Int, Array{Int, 1}},
...
-#         for neighbor in graph.neighbors[next_node]
-#             neighbor_edge = graph.adj_matrix[neighbor, next_node]
-#             if (neighbor_edge in mst_edges && !(neighbor in traversed_nodes) && neighbor != avoid_node)
-#                 push!(stack, neighbor)
-#             end
-#         end
-#     end
-#     return traversed_nodes
-# end
-
```
Comment (bsuwal) [inline src/recom.jl]: Cleanup

File: src/recom.jl [hunk 5]
```diff
@@ -176,9 +153,9 @@ function get_valid_proposal(graph::BaseGraph,
 end
 
 
-function update!(graph::BaseGraph,
-                 partition::Partition,
-                 proposal::RecomProposal)
+function update_partition!(partition::Partition,
+                           graph::BaseGraph,
+                           proposal::RecomProposal)
```
Comment (bsuwal) [inline src/recom.jl]: update!() was a bad name

File: test/make_4x4_test_grid.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+g.to_json("test_grid_4x4.json")
+
+# color by assignment
+plt.figure()
+nx.draw(
+    graph,
+    # pos={x: x for x in node_coords},
+    node_color=[graph.nodes[x]["assignment"] for x in graph.nodes()]
+)
+plt.show()
```
Comment (bsuwal) [inline test/make_4x4_test_grid.py]: Decided to put in the file that generates the test 4x4 grid

File: src/GerryChain.jl [hunk 7]
```diff
@@ -23,13 +23,18 @@ satisfy_constraint,
 
 # recom
 sample_subgraph, build_mst, add_edge_to_mst!, remove_edge_from_mst!,
-traverse_mst, get_balanced_proposal, get_valid_proposal, update!,
-recom_chain
+traverse_mst, get_balanced_proposal, get_valid_proposal, update_partition!,
+recom_chain,
+
+# measures
+AbstractMeasure, Election, RacePopulation,
```
Comment (pjrule) [inline src/GerryChain.jl:30]: Lol, the first thing that comes to mind is [a measure in the analysis sense.](https://en.wikipedia.org/wiki/Measure_(mathematics)

File: test/make_4x4_test_grid.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        graph.nodes[i]["assignment"] = 4
+
+def set_election_data(graph):
+    for i in (0, 1, 2, 3, 8, 9, 10, 11):
+        graph.nodes[i]["electionD"] = 2
+        graph.nodes[i]["electionR"] = 0
+    for j in (4, 5, 6, 7, 12, 13, 14, 15):
+        graph.nodes[j]["electionD"] = 1
+        graph.nodes[j]["electionR"] = 3
+
```
Comment (pjrule) [inline test/make_4x4_test_grid.py:28]: can this be more compact?
```
purple_pops = [...]
pink_pops = [...]
for node, purple, pink in zip(graph.nodes, purple_pops, pink_pops):
    graph.nodes[node]['purple'] = purple
    graph.nodes[node]['pink'] = pink
```

**Pull request merged**